test/test_provider_collector.py

```diff
@@ -0,0 +1,34 @@
+# coding: utf-8
+
+import eventlet
+import pytest
+
+from nameko.extensions import ProviderCollector
+
+
+def test_provider_collector():
+    collector = ProviderCollector()
+
+    provider1 = object()
+    provider2 = object()
+    collector.register_provider(provider1)
+    collector.register_provider(provider2)
+
+    assert provider1 in collector._providers
+    assert provider2 in collector._providers
+
+    collector.unregister_provider(provider1)
+    assert provider1 not in collector._providers
+
+    # unregister missing provider is a no-op
+    collector.unregister_provider(provider1)
+
+    # stop() should block while a provider remains
+    with pytest.raises(eventlet.Timeout):
+        with eventlet.Timeout(0):
```

Comment:
``collect.stop`` doesn't yield, so this is fine. The ``ProviderCollector`` is slated for deprecation anyway...